metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -1267,8 +1275,21 @@ class MetalsLspService(
           .startsWith(reports.bloop.maybeReportsDir)
       )
     if (bloopReportDelete.nonEmpty) connectionBspStatus.onReportsUpdate()
-    otherDeleteEvents.map(_.getUri().toAbsolutePath).foreach(onDelete)
-    onChange(changeAndCreateEvents.map(_.getUri().toAbsolutePath))
+    def toPath(event: FileEvent) = event.getUri().toAbsolutePath
+    otherDeleteEvents.map(toPath).foreach(onDelete)
+    // when metals creates `.bsp/<name>.json` files we want to ignore
+    // create event for those files
+    val ignoreJsonCreation = expectBspConfigCreation.get
```

Comment:
Do we need this change? Could we just not connect to the server created if we are currently connecting already? So maybe ignore things in onBuildChangedUnbatched ?

Reply:
> Could we just not connect to the server created if we are currently connecting already? 

Yes and that is probably the proper fix want to have. The tricky part is that we need to be able to `cancel` if it gets stuck. I think we have some timeouts but I'm not sure on which parts and still the user might prefer to switch build server right away instead of waiting for the timeout. In that case we should cancel the old connection attempt. (https://github.com/scalameta/metals/issues/2563)
> So maybe ignore things in onBuildChangedUnbatched ?

That file is created before we start connecting. I guess we can ignore things in `onBuildChangedUnbatched` from before creating `json` until we finish connecting. But then if in the meantime someone creates `build.sbt` it won't suggest connecting to `sbt` at all. Let me see if I can come up with a better way to simplify this.

Anyway this PR is not a proper fix it just deduplicates things for `scala-cli`  but it doesn't resolve the underlying issue. I'd still be pro pushing a workaround, 'cause for `scala-cli` it's very visibly causing issues now and creating a proper fix might be a bit more complex.

Reply:
I ended up ignoring everything in `onBuildChangedUnbatched` until finishes connecting/reconnecting.